frame/mosaic-vault/src/lib.rs

```diff
@@ -0,0 +1,904 @@
+            <Deposits<T>>::insert(asset_id, DepositInfo{asset_id, amount});
+
+			Self::deposit_event(Event::DepositCompleted{
+				sender,
+				asset_id,
+				remote_asset_id: Self::remote_asset_id(remote_network_id, asset_id),
+				remote_network_id,
+				destination_address,
+				amount, 
+				deposit_id,
+				transfer_delay
+			});
+
+			Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn withdraw(
+			origin: OriginFor<T>, 
+			destination_account: T::AccountId,
+			amount: T::Balance,
+			asset_id: T::AssetId, 
+			remote_network_id: T::RemoteNetworkId,
+	        deposit_id: T::DepositId,
+			fee: T::Balance,
+		 ) -> DispatchResultWithPostInfo {
+
+			 let sender = ensure_signed(origin.clone())?;
+
+			 T::RelayerOrigin::ensure_origin(origin)?;
+         
+			 ensure!(Self::pause_status() == false, Error::<T>::ContractPaused);
+
+			 Self::only_supported_remote_token(remote_network_id.clone(), asset_id.clone())?;
+             
+			ensure!(Self::has_been_withdrawn(deposit_id) == false, Error::<T>::AlreadyWithdrawn);
+
+			ensure!(Self::get_current_token_liquidity(asset_id)? >= amount, Error::<T>::InsufficientAssetBalance); 
+
+			  <HasBeenWithdrawn<T>>::insert(deposit_id, true);
+
+			  <LastWithdrawID<T>>::put(deposit_id);
+
+			  let pallet_account_id = Self::account_id(); 
+	
+			  let withdraw_amount = amount.saturating_sub(fee);
+
+			 T::Currency::mint_into(asset_id, &pallet_account_id, amount).map_err(|_|Error::<T>::MintToFailed)?;
+
+			 Self::decrease_total_value_transferred(asset_id, amount)?;
+
+			  T::Currency::transfer(asset_id, &pallet_account_id, &destination_account, withdraw_amount, true).map_err(|_|Error::<T>::TransferFromFailed)?;
+			
+			 if fee > T::Balance::zero() {  
+			   
+				 T::Currency::transfer(asset_id, &pallet_account_id, &Self::get_fee_address(), fee, true).map_err(|_|Error::<T>::TransferFromFailed)?;
+				
+				Self::deposit_event(Event::FeeTaken{
+					sender, 
+					destination_account: destination_account.clone(), 
+					asset_id,
+					amount,
+					fee,
+					deposit_id,
+				});
+			 }
+
+			 Self::deposit_event(Event::WithdrawalCompleted{
+				destination_account,
+				amount,
+				withdraw_amount,
+				fee,
+				asset_id,
+				deposit_id
+			 });
+
+			 Ok(().into())
+		 }
+
+		 #[pallet::weight(10_000)]
+		 pub fn unlock_in_transfer_funds(
+			origin: OriginFor<T>,
+			asset_id: T:: AssetId,
+			amount: T::Balance,
+			deposit_id: T::DepositId,
+		 ) ->DispatchResultWithPostInfo {
+
+			T::RelayerOrigin::ensure_origin(origin)?;
+
+			ensure!(Self::pause_status() == false, Error::<T>::ContractPaused);
+			
+			ensure!(Self::has_been_completed(deposit_id) == false, Error::<T>::AlreadCompleted);
+
+			ensure!(Self::in_transfer_funds(asset_id) >= amount, Error::<T>::InsufficientFunds);
+
+			let deposit = Self::deposits(asset_id);
+
+			ensure!(deposit.asset_id == asset_id && deposit.amount == amount, Error::<T>::InsufficientFunds);
```

Comment:
not sure about deposit. is it share for asset? if not, and created each time as its id, than why to pass amount? anyway check for ==

Reply:
It keeps track of deposits , once the transfer is completed by the relayer, it is flagged as complete,